### ISSUE TITLE
Support plugins. Decouple functionalities by an event driven architecture

### ISSUE THREAD

**Issue by megastef** [sematext/logagent-js#28]
In logagent 1.x the CLI/server functionalities are tied together in a large script. In 2.x we started to split it into multiple files. The goal is to have plugins, like Etsy statsd backends, which could be installed as separate npm modules to extend the functionality. 

How to do this? 
- Introduce a central event queue with topics and pub/sub (a bit like Kafka, but in-process). 
  eventemitter2 seems to be a good package for this approach. So plugins could work together without having dependencies to each other. This should be wrapped in a separate "class", so we could support in the future multi-process/IPC plugins by exchanging the central event router (e.g. using e.g. MQTT for distributed workloads). 
- each event must include metadata as context (e.g. source of logs, or Elasticsearch index to use, ...)
- A simple plugin interface giving config to constructor and having / start(), stop() functions to register for events. Plugins should emit events. In addition plugins should be able to expose CLI or at least config file options. 

What to do? 
- rewrite exiting inputs/outputs for the event driven architecture (POC done locally for stdin/stdout/elasticsearch)
- implement plugin loader mechanism (and unload on SIGTERM/SIGINT/SIGQUIT)...
- implement config loader interface, so plugins could expose available options to the config loader

**Comment by megastef** [comment]
Example Plugins: 
https://github.com/megastef/logagent-tcp-input

**Comment by otisg** [comment]
Just curious:
* if N outputs are defined, is data sent to all of them in parallel or sequentially or undefined?
* can one control / filter which data goes to which output by either mapping 1 or more inputs to 1 or more outputs, or by some sort of a data matching filter?

**Comment by megastef** [comment]
@otisg - Currently all N outputs receive the processed input.
